src/app/workflow/dag/cwl-viewer/cwl-viewer.service.ts

```diff
@@ -0,0 +1,119 @@
+
+interface WorkflowDetails {
+  /**
+   * A relative path link to the svg format visualisation image
+   */
+  visualisationSvg: string;
+}
+
+/**
+ * A descriptor for some of the urls available from view.commonwl.org for a CWL visualization.
+ */
+export interface CwlViewerDescriptor {
+  /**
+   * The url of the SVG file
+   */
+  svgUrl: string;
+  /**
+   * The url of the web page on view.commonwl.org
+   */
+  webPageUrl: string;
+}
+
+/**
+ * This service is responsible for generating visualizations, if they haven't already been generated, from
+ * https://view.commonwl.org, and returning the urls of those visualizations.
+ */
+@Injectable()
+export class CwlViewerService {
+
+  constructor(private httpClient: HttpClient) {
+  }
+
+  /**
+   *
+   * Returns visualization urls for a workflow.
+   *
+   * 1. POSTs to https://view.commonwl.org/workflows?url=<cwl on GitHub>
+   *    1. If visualization has already been generated, or generation is quick, response is 200 with details in response body.
+   *    Actually, the POST returns a 303, but Angular HttpClient follows the redirect so from the client perspective
+   *    it's as if it returned a 200.
+   *    2. If image generation needs to be queued up, response is 202 with link to job in Location header
+   *      1. Follow location header and poll job queue
+   *      2. When job is finished, polling job queue redirects and returns a response with details in body
+   *
+   * @param {string} providerUrl, e.g., `https://github.com/dockstore-testing/Metaphlan-ISBCGC`
+   * @param {string} reference, a branch name, e.g., `master`
+   * @param {string} workflow_path, path to CWL, e.g., `/metaphlan_wfl.cwl`
+   * @returns {Observable<CwlViewerDescriptor>}
+   */
+  getVisualizationUrls(providerUrl: string, reference: string, workflow_path: string): Observable<CwlViewerDescriptor> {
+
+    const url = this.cwlViewerEndpoint(providerUrl, reference, workflow_path);
+
+    return this.httpClient.post(url, null, {observe: 'response'})
+      .switchMap((res: HttpResponse<WorkflowDetails>) => {
+        if (res.status === 200) {
+          return Observable.of(<CwlViewerDescriptor>{
+            svgUrl: Dockstore.CWL_VISUALIZER_URI + res.body.visualisationSvg,
+            webPageUrl: res.url});
+        } else if (res.status === 202) {
+          const locationHeader = res.headers.get('Location');
+          if (locationHeader) {
+            const queueUrl = locationHeader.startsWith('/') ? Dockstore.CWL_VISUALIZER_URI + locationHeader : locationHeader;
+            return this.pollJobQueue(queueUrl);
+          }
+        }
+        throw new Error(`Error posting ${workflow_path}`);
+      });
+  }
+
+  /**
+   * Returns the CommonWL endpoint to request a visualization.
+   *
+   * @param {string} providerUrl
+   * @param {string} reference
+   * @param {string} workflow_path
+   * @returns {string}
+   */
+  cwlViewerEndpoint(providerUrl: string, reference: string, workflow_path: string) {
```

Comment:
Doesn't actually have a string return type defined `: string`

Reply:
Was more of an observation that the jsdoc has `@returns {string}` but the method doesn't explicitly define it.  My jsdocs are auto-generated from the method signature, so I find it odd that there's a mismatch